File: 0001-0500/0111_minDepth.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int minDepth(TreeNode root) {
        if (root == null) return 0;

        Queue<TreeNode> queue = new LinkedList<>() {{add(root);}};
        int depth = 0;

        while (!queue.isEmpty()) {
            int length = queue.size();
            depth++;

            for (int i = 0; i < length; i++) {
                TreeNode node = queue.remove();
                if (node.left == null && node.right == null) return depth;
                if (node.left != null) queue.add(node.left);
                if (node.right != null) queue.add(node.right);
            }
        }

        return -1;
    }
}

File: 0111_minDepth.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def __init__(self):
        self.min = float('inf')
        
    def minDepth(self, root: Optional[TreeNode]) -> int:
        if not root: return 0
        height = 1
        self.dfs(root, height)
        return self.min
        
    def dfs(self, node, height):
        if not node.left and not node.right:
            self.min = min(self.min, height)
            return
        if node.left:
            self.dfs(node.left, height+1)
        if node.right:
            self.dfs(node.right, height+1)

# Alternative solution

class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        queue = deque([root])
        depth = 0
        while queue:
            levelSize = len(queue)
            depth += 1
            for _ in range(levelSize):
                node = queue.popleft()
                if not node.left and not node.right:
                    return depth
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
        return depth
